sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -771,44 +781,67 @@ export class MessageReceiver extends LinkEntity {
           this.name = options.name;
         }
 
-        this.isConnecting = true;
-        await this._negotiateClaim();
-        if (!options) {
-          options = this._createReceiverOptions();
-        }
-        log.error(
-          "[%s] Trying to create receiver '%s' with options %O",
-          connectionId,
-          this.name,
-          options
-        );
+        await defaultLock.acquire(this._openLock, async () => {
+          if (this.wasCloseInitiated) {
+            const err = new MessagingError(
```

Comment:
Does this have to be a `MessagingError` instead of a simple error? Generally, we save this class for errors that might be related with us talking to the service and not much of how the user interacts with the client.

Is the reason for having this so that the retry() logic in `onDetached` considers this as not retryable?
Wouldnt that need be served if threw no error at all?

Reply:
Two parts:
1. It's definitely a fatal error for this code to be called if the client is already closed. How the caller handles it is definitely up to them but I wanted this code to properly protect itself.
2. I can definitely change the error type - it just seemed to be the common internal one. Do we just typically instantiate an error and then set .retryable on it then?

Reply:
Changed over to use an Error (basically).

Reply:
I am still not sold that we should throw an error from here.

Errors for when a receiver is used in any which way after it is closed are thrown from the outer `Receiver` close currently.

Even for the streaming receiver, we do check the `isClosed` flag on the outside after `init()` is resolved. See https://github.com/Azure/azure-sdk-for-js/blob/%40azure/service-bus_1.1.6/sdk/servicebus/service-bus/src/receiver.ts#L141

I would recommend that we treat init() as a no-op here and throw the error from the outside where you can re-use the `throwIfReceiverOrConnectionClosed` helper method.

Reply:
Another reason for not throwing an error from here is the case of link recovery.
Take the case of user closing the receiver when we are in the midst of recovering it.
This change will result in the error being thrown for no fault of the user

Reply:
On second thoughts, I am concerned about throwing this error even from the outside, atleast for Track 1. Patch updates arent meant to introduce new error scenarios. This error will bubble up via the user's error handler which till now did not have to deal with the receiver is closed error.

Reply:
Great points, changed to just be a simple return.